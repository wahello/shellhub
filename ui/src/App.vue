<template>
  <v-app>
    <Snackbar />

    <v-navigation-drawer
      v-if="isLoggedIn && hasNamespaces"
      v-model="showNavigationDrawer"
      app
      clipped
      dark
    >
      <v-list>
        <v-list-item
          v-for="item in visibleItems"
          :key="item.title"
          :to="item.path"
          two-line
        >
          <v-list-item-action>
            <v-icon v-text="item.icon" />
          </v-list-item-action>

          <v-list-item-content>
            <v-list-item-title
              :data-test="item.icon"
              v-text="item.title"
            />
          </v-list-item-content>
        </v-list-item>
      </v-list>
    </v-navigation-drawer>

    <AppBar
      :drawer.sync="drawer"
    />

    <router-view
      v-if="!isLoggedIn"
      key="/login"
    />

    <v-main
      v-if="isLoggedIn"
      class="grey lighten-4"
    >
      <v-container
        class="pa-8"
        fluid
      >
        <router-view :key="$route.fullPath" />
      </v-container>

      <snackbar />
    </v-main>

    <v-overlay :value="hasSpinner">
      <v-progress-circular
        indeterminate
        size="64"
      />
    </v-overlay>

    <DeviceWarning data-test="deviceWarning-component" />
  </v-app>
</template>

<script>

import Snackbar from '@/components/snackbar/Snackbar';
import AppBar from '@/components/app_bar/AppBar';
import DeviceWarning from '@/components/device/DeviceWarning';

export default {
  name: 'App',

  components: {
    Snackbar,
    AppBar,
    DeviceWarning,
  },

  data() {
    return {
      drawer: false,
      clipped: false,
      items: [
        {
          icon: 'dashboard',
          title: 'Dashboard',
          path: '/',
        },
        {
          icon: 'devices',
          title: 'Devices',
          path: '/devices',
        },
        {
          icon: 'history',
          title: 'Sessions',
          path: '/sessions',
        },
        {
          icon: 'security',
          title: 'Firewall Rules',
          path: '/firewall/rules',
          hidden: !this.$env.isEnterprise,
        },
        {
          icon: 'vpn_key',
          title: 'Public Keys',
          path: '/sshkeys/public-keys',
        },
      ],
      admins: [
        ['Management', 'people_outline'],
        ['Settings', 'settings'],
      ],
    };
  },

  computed: {
    isLoggedIn() {
      return this.$store.getters['auth/isLoggedIn'];
    },

    visibleItems() {
      return this.items.filter((item) => !item.hidden);
    },

    hasLoggedID() {
      return this.$store.getters['auth/id'] !== '';
    },

    hasNamespaces() {
      return this.$store.getters['namespaces/getNumberNamespaces'] !== 0;
    },

    hasSpinner() {
      return this.$store.getters['spinner/getStatus'];
    },

    showNavigationDrawer: {
      get() {
        return !this.$store.getters['mobile/isMobile'] || this.drawer;
      },

      set(status) {
        this.drawer = status;
      },
    },
  },

  beforeDestroy() {
    if (typeof window === 'undefined') return;

    window.removeEventListener('resize', this.onResize, { passive: true });
  },

  created() { // previous user tenant changed into a namespace
    if (!this.hasLoggedID && this.isLoggedIn) {
      try {
        this.$store.dispatch('auth/logout').then(() => {
          this.$router.push('/login');
        });
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceReload);
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$error.namespaceLoad);
      }
    }
  },

  mounted() {
    this.onResize();
    window.addEventListener('resize', this.onResize, { passive: true });

    this.$store.dispatch('privatekeys/fetch');
  },

  methods: {
    triggerClick(item) {
      switch (item.type) {
      case 'path':
        this.$router.push(item.path).catch(() => {});
        break;
      case 'method':
        this[item.method]();
        break;
      default:
        break;
      }
    },

    onResize() {
      const isMobile = this.$vuetify.breakpoint.mobile;
      this.$store.dispatch('mobile/setIsMobileStatus', isMobile);
    },
  },
};

</script>

<style>
.v-list-active {
  border-left: 4px solid var(--v-primary-base);
}

.text-shadow {
  text-shadow: #000 0 0 6px;
  color: transparent;
}
</style>
