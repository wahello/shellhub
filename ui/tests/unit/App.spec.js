import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import Router from 'vue-router';
import App from '@/App';

describe('App', () => {
  const localVue = createLocalVue();
  const router = new Router();
  localVue.use(Vuex);
  localVue.use(Router);
  const vuetify = new Vuetify();

  let wrapper;

  const tenant = 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx';
  const isLoggedIn = true;
  const isMobile = false;
  const hasSpinner = false;

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      tenant,
      isLoggedIn,
      isMobile,
      hasSpinner,
    },
    getters: {
      'auth/tenant': (state) => state.tenant,
      'auth/isLoggedIn': (state) => state.isLoggedIn,
      'mobile/isMobile': (state) => state.isMobile,
      'spinner/getStatus': (state) => state.hasSpinner,
    },
    actions: {
      'auth/logout': () => {},
      'privatekeys/fetch': () => {},
      'mobile/setIsMobileStatus': () => {},
    },
  });

  beforeEach(() => {
    wrapper = shallowMount(App, {
      store,
      localVue,
      stubs: ['fragment'],
      mocks: {
        $env: (isEnterprise) => isEnterprise,
      },
      router,
      vuetify,
    });
  });

  it('Is a Vue instance', () => {
    expect(wrapper).toBeTruthy();
  });
  it('Renders the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('Renders the template with components', async () => {
    expect(wrapper.find('[data-test="deviceWarning-component"]').exists()).toEqual(true);
  });
  it('Renders the template with data', async () => {
    expect(wrapper.find('[data-test="dashboard"]').exists()).toEqual(true);
    expect(wrapper.find('[data-test="devices"]').exists()).toEqual(true);
    expect(wrapper.find('[data-test="history"]').exists()).toEqual(true);
  });
});
