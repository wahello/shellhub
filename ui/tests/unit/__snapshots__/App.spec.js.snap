// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders the component 1`] = `
<v-app-stub id="app">
  <snackbar-stub></snackbar-stub>
  <v-navigation-drawer-stub app="true" mobilebreakpoint="1264" dark="true" clipped="true" height="100vh" minivariantwidth="56" src="" tag="nav" width="256" value="true">
    <v-list-stub tag="div">
      <v-list-item-stub activeclass="" to="/" tag="div" twoline="true">
        <v-list-item-action-stub>
          <v-icon-stub>dashboard</v-icon-stub>
        </v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub datatest="dashboard" data-test="dashboard">Dashboard</v-list-item-title-stub>
        </v-list-item-content-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" to="/devices" tag="div" twoline="true">
        <v-list-item-action-stub>
          <v-icon-stub>devices</v-icon-stub>
        </v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub datatest="devices" data-test="devices">Devices</v-list-item-title-stub>
        </v-list-item-content-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" to="/sessions" tag="div" twoline="true">
        <v-list-item-action-stub>
          <v-icon-stub>history</v-icon-stub>
        </v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub datatest="history" data-test="history">Sessions</v-list-item-title-stub>
        </v-list-item-content-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" to="/sshkeys/public-keys" tag="div" twoline="true">
        <v-list-item-action-stub>
          <v-icon-stub>vpn_key</v-icon-stub>
        </v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub datatest="vpn_key" data-test="vpn_key">Public Keys</v-list-item-title-stub>
        </v-list-item-content-stub>
      </v-list-item-stub>
    </v-list-stub>
  </v-navigation-drawer-stub>
  <appbar-stub></appbar-stub>
  <!---->
  <v-main-stub tag="main" class="grey lighten-4">
    <v-container-stub tag="div" fluid="true" class="pa-8">
      <router-view-stub name="default"></router-view-stub>
    </v-container-stub>
    <snackbar-stub></snackbar-stub>
  </v-main-stub>
  <v-overlay-stub color="#212121" dark="true" opacity="0.46" zindex="5">
    <v-progress-circular-stub indeterminate="true" rotate="0" size="64" width="4" value="0"></v-progress-circular-stub>
  </v-overlay-stub>
  <devicewarning-stub data-test="deviceWarning-component"></devicewarning-stub>
</v-app-stub>
`;
